src/main/java/jcifs/smb/SmbTransportPoolImpl.java

```diff
@@ -0,0 +1,355 @@
+                            log.trace("Cannot reuse, different config " + conn);
+                        }
+                        continue;
+                    }
+                }
+                catch ( CIFSException e ) {
+                    log.debug("Error while checking for reuse", e);
+                    continue;
+                }
+
+                if ( log.isTraceEnabled() ) {
+                    log.trace("Reusing transport connection " + conn);
+                }
+                return conn.acquire();
+            }
+        }
+
+        return null;
+    }
+
+
+    @Override
+    public SmbTransportImpl getSmbTransport ( CIFSContext tf, String name, int port, boolean exclusive, boolean forceSigning ) throws IOException {
+
+        Address[] addrs = tf.getNameServiceClient().getAllByName(name, true);
+
+        if ( addrs == null || addrs.length == 0 ) {
+            throw new UnknownHostException(name);
+        }
+
+        Arrays.sort(addrs, new Comparator<Address>() {
+
+            @Override
+            public int compare ( Address o1, Address o2 ) {
+                Integer fail1 = SmbTransportPoolImpl.this.failCounts.get(o1.getHostAddress());
+                Integer fail2 = SmbTransportPoolImpl.this.failCounts.get(o2.getHostAddress());
+                if ( fail1 == null ) {
+                    fail1 = 0;
+                }
+                if ( fail2 == null ) {
+                    fail2 = 0;
+                }
+                return Integer.compare(fail1, fail2);
+            }
+
+        });
+
+        synchronized ( this.connections ) {
+            for ( Address addr : addrs ) {
+                SmbTransportImpl found = findConnection(
+                    tf,
+                    addr,
+                    port,
+                    tf.getConfig().getLocalAddr(),
+                    tf.getConfig().getLocalPort(),
+                    name,
+                    forceSigning,
+                    true);
+                if ( found != null ) {
+                    return found;
+                }
+            }
+        }
+
+        IOException ex = null;
+        for ( Address addr : addrs ) {
+            try ( SmbTransportImpl trans = getSmbTransport(tf, addr, port, exclusive, forceSigning).unwrap(SmbTransportImpl.class) ) {
+                try {
+                    trans.ensureConnected();
+                }
+                catch ( IOException e ) {
+                    removeTransport(trans);
+                    throw e;
+                }
+                return trans.acquire();
+            }
+            catch ( IOException e ) {
+                String hostAddress = addr.getHostAddress();
+                Integer failCount = this.failCounts.get(hostAddress);
+                if ( failCount == null ) {
+                    this.failCounts.put(hostAddress, 1);
+                }
+                else {
+                    this.failCounts.put(hostAddress, failCount + 1);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **227, 228, and 232**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).